providers/kind/kind.sh

```diff
@@ -0,0 +1,173 @@
+# This file is part of the Kepler project
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Copyright 2022 The Kepler Contributors
+#
+
+set -eu -o pipefail
+
+declare -r KIND_DIR="${KIND_DIR:-"$PROJECT_ROOT/tmp/kind"}"
+declare -r KIND="$BIN_DIR/kind"
+declare -r KIND_VERSION=${KIND_VERSION:-0.17.0}
+declare -r KIND_CONFIG_YAML="$KIND_DIR/kind.yml"
+
+declare -r KIND_REGISTRY_NAME=${KIND_REGISTRY_NAME:-kind-registry}
+declare -r KIND_REGISTRY_YAML="$KIND_DIR/local-registry.yml"
+
+declare -r KIND_DEFAULT_NETWORK="kind"
+declare -r KIND_CONFIG_PATH="$PROJECT_ROOT/providers/kind"
+declare -r KIND_MANIFESTS_DIR="$KIND_CONFIG_PATH/manifests"
+declare -r KIND_CLUSTER_NAME=${KIND_CLUSTER_NAME:-kind}
+declare -r KIND_IMAGE_REPO=${KIND_IMAGE_REPO:-localhost:5001}
+
+# check CPU arch
+cpu_arch() {
+	case "$(uname -m)" in
+	x86_64* | i?86_64* | amd64*)
+		echo "amd64"
+		;;
+	ppc64le)
+		echo "ppc64le"
+		;;
+	aarch64* | arm64*)
+		echo "arm64"
+		;;
+	*)
+		return 1
+		;;
+	esac
+	return 0
+}
+
+kind_install() {
+	if [[ -f "$KIND" ]] &&
+		[[ $($KIND --version | awk '{print $3}') == "$KIND_VERSION" ]]; then
+		info "kind already installed; skipping installation"
+		return 0
+	fi
+
+	local arch
+	arch="$(cpu_arch)"
+	[[ "$arch" == "" ]] && {
+		err "failed to determine CPU arch; cannot setup kind"
+		return 1
+	}
+
+	local os="linux"
+	[[ "$OSTYPE" =~ darwin ]] && os=darwin
+
+	info "Downloading kind v$KIND_VERSION for $os - $arch"
+	curl -LSs https://github.com/kubernetes-sigs/kind/releases/download/v$KIND_VERSION/kind-${os}-${arch} -o "$KIND"
+	chmod +x "$KIND"
+}
+
+kind_wait_up() {
+	echo "Waiting for kind to be ready ..."
+
+	while [ -z "$($CTR_CMD exec --privileged "${KIND_CLUSTER_NAME}"-control-plane kubectl --kubeconfig=/etc/kubernetes/admin.conf get nodes -o=jsonpath='{.items..status.conditions[-1:].status}' | grep True)" ]; do
+		echo "Waiting for kind to be ready ..."
+		sleep 10
+	done
+	echo "Waiting for dns to be ready ..."
+	kubectl wait -n kube-system --timeout=12m --for=condition=Ready -l k8s-app=kube-dns pods
+}
+#
+_prepare_config() {
+	local default_registry_name="kind-registry"
+	local default_registry_port="5001"
+
+	info "Building manifests..."
+	cp $KIND_MANIFESTS_DIR/kind.yml "$KIND_CONFIG_YAML"
+
+	sed "s/$default_registry_name/${KIND_REGISTRY_NAME}/g" "$KIND_CONFIG_YAML" >"${KIND_CONFIG_YAML}.tmp" &&
```

Comment:
I never liked this sed implementation. I was doing some exploration on [this](https://github.com/sustainable-computing-io/local-dev-cluster/pull/28#discussion_r1258171819) and came across `ex` Maybe we can make this change. Wdyt?

Reply:
It seems that `sed -i` shows better readability :-)

Reply:
so `sed` has different implementations on Linux and macOS. On linux `sed -i -e` works fine but on macOS it creates `-e` backup files. When this [step](https://github.com/sustainable-computing-io/local-dev-cluster/blob/e7bd1d061e6806ea91c1caf9448d5e24c292e55d/prometheus.sh#L43) is triggered using `sed` it creates an issue as kubectl tries to deploy backup file as well and at that point the script breaks because the modified file got already deployed earlier. Here is related [PR](https://github.com/sustainable-computing-io/local-dev-cluster/pull/9). That's why I went with the current implementation of redirecting output to tmp file and then mv it.

The other workaround to this which I found was to use `ex` and I think the implementation is same on Linux and macOS

Reply:
@jiere , indeed, I was thinking the same but I was sure there must have been a reason :D. Given that the file is copied, I think we can `cat <orig> | sed  > generated.yaml` instead of `-i' . In any case, I didn't want to change what's working currently too much (beyond refactoring).